File: projects/android/Sample/app/src/main/java/com/ezored/sample/app/Application.java
package com.ezored.sample.app;

import android.arch.lifecycle.ProcessLifecycleOwner;
import android.content.IntentFilter;
import android.support.multidex.MultiDexApplication;
import android.support.v7.app.AppCompatDelegate;
import android.util.Log;

import com.dropbox.djinni.JNILoader;
import com.ezored.core.ApplicationCore;
import com.ezored.data.SharedData;
import com.ezored.data.SharedDataPlatformServiceImpl;
import com.ezored.domain.DeviceData;
import com.ezored.domain.InitializationData;
import com.ezored.io.FileHelper;
import com.ezored.io.FileHelperPlatformServiceImpl;
import com.ezored.sample.BuildConfig;
import com.ezored.sample.data.AppData;
import com.ezored.sample.observers.AppLifecycleObserver;
import com.ezored.sample.receivers.ConnectivityChangeReceiver;
import com.ezored.sample.utils.EnvironmentUtil;
import com.ezored.sample.utils.Utils;
import com.ezored.util.Logger;
import com.ezored.util.LoggerLevel;
import com.ezored.util.LoggerPlatformServiceImpl;

public class Application extends MultiDexApplication {

    private static String LOG_GROUP = "SAMPLE";
    private static Application instance;
    private AppData appData;

    @Override
    public void onCreate() {
        super.onCreate();

        instance = this;

        appData = new AppData();

        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);

        loadNativeLibrary();

        ProcessLifecycleOwner.get().getLifecycle().addObserver(new AppLifecycleObserver());

        registerReceivers();

        if (BuildConfig.DEBUG) {
            Utils.showFacebookKeyHash(this);
        }
    }

    public AppData getAppData() {
        return appData;
    }

    private void registerReceivers() {
        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction("android.net.conn.CONNECTIVITY_CHANGE");

        registerReceiver(new ConnectivityChangeReceiver(), intentFilter);
    }

    public static Application getInstance() {
        return instance;
    }

    @Override
    public void onTerminate() {
        Logger.i("[Application : onTerminate] App terminated");
        super.onTerminate();
    }

    public void onMoveToForeground() {
        Logger.i("[Application : onMoveToForeground] App moved to foreground");
        getAppData().setAppInBackground(false);
    }

    public void onMoveToBackground() {
        Logger.i("[Application : onMoveToBackground] App moved to background");
        getAppData().setAppInBackground(true);
    }

    private void loadNativeLibrary() {
        try {
            System.loadLibrary("Ezored");

            JNILoader.load();

            initializeLogger();
            initializeSharedData();
            initializeFileHelper();
            initializeCore();
        } catch (UnsatisfiedLinkError e) {
            Log.e(LOG_GROUP, "Could not load native library");
            e.printStackTrace();
        }
    }

    private void initializeCore() {
        InitializationData initializationData = EnvironmentUtil.getInitializationData();
        DeviceData deviceData = EnvironmentUtil.getDeviceData();

        ApplicationCore.shared().initialize(initializationData, deviceData);
    }

    private void initializeLogger() {
        Logger.shared().setPlatformService(new LoggerPlatformServiceImpl(LOG_GROUP));

        if (BuildConfig.DEBUG) {
            Logger.shared().setLevel(LoggerLevel.VERBOSE);
        } else {
            Logger.shared().setLevel(LoggerLevel.ERROR);
        }
    }

    private void initializeSharedData() {
        SharedData.shared().setPlatformService(new SharedDataPlatformServiceImpl(getApplicationContext()));
    }

    private void initializeFileHelper() {
        FileHelper.shared().setPlatformService(new FileHelperPlatformServiceImpl());
    }

}

File: projects/android/Sample/app/src/main/java/com/ezored/sample/app/Application.kt
package com.ezored.sample.app

import android.content.IntentFilter
import android.content.res.AssetManager
import android.os.StrictMode
import android.util.Log
import androidx.appcompat.app.AppCompatDelegate
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.multidex.MultiDexApplication
import com.dropbox.djinni.JNILoader
import com.ezored.core.ApplicationCore
import com.ezored.data.SharedData
import com.ezored.data.SharedDataPlatformServiceImpl
import com.ezored.io.FileHelper
import com.ezored.io.FileHelperPlatformServiceImpl
import com.ezored.net.http.HttpClient
import com.ezored.net.http.HttpClientPlatformServiceImpl
import com.ezored.net.http.HttpServer
import com.ezored.net.http.HttpServerConfig
import com.ezored.sample.BuildConfig
import com.ezored.sample.data.AppData
import com.ezored.sample.extension.copyAssetFolder
import com.ezored.sample.observer.AppLifecycleObserver
import com.ezored.sample.receiver.ConnectivityChangeReceiver
import com.ezored.sample.util.EnvironmentUtil
import com.ezored.util.Logger
import com.ezored.util.LoggerLevel
import com.ezored.util.LoggerPlatformServiceImpl
import com.google.android.instantapps.InstantApps

class Application : MultiDexApplication() {

    var appData: AppData = AppData()

    override fun onCreate() {
        super.onCreate()

        instance = this

        initializeStrictMode()

        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)

        checkInstantApp()

        loadNativeLibrary()

        ProcessLifecycleOwner.get().lifecycle.addObserver(AppLifecycleObserver())

        registerReceivers()
    }

    private fun checkInstantApp() {
        if (InstantApps.isInstantApp(this)) {
            appData.isInstantApp = true
        }
    }

    private fun registerReceivers() {
        Logger.i("[Application : registerReceivers]")

        val intentFilter = IntentFilter()
        intentFilter.addAction("android.net.conn.CONNECTIVITY_CHANGE")

        registerReceiver(ConnectivityChangeReceiver(), intentFilter)
    }

    override fun onTerminate() {
        Logger.i("[Application : onTerminate] App terminated")
        super.onTerminate()
    }

    fun onMoveToForeground() {
        Logger.i("[Application : onMoveToForeground] App moved to foreground")
        appData.isAppInBackground = false
    }

    fun onMoveToBackground() {
        Logger.i("[Application : onMoveToBackground] App moved to background")
        appData.isAppInBackground = true
    }

    private fun loadNativeLibrary() {
        try {
            System.loadLibrary("ezored")

            JNILoader.load()

            initializeLogger()
            initializeSharedData()
            initializeHttpClient()
            initializeFileHelper()
            initializeCore()
            initializeHttpServer()
        } catch (e: UnsatisfiedLinkError) {
            Log.e(LOG_GROUP, "Could not load native library: " + e.message)
            e.printStackTrace()
        }
    }

    private fun initializeCore() {
        val initializationData = EnvironmentUtil.initializationData
        val deviceData = EnvironmentUtil.deviceData

        ApplicationCore.shared().initialize(initializationData, deviceData)
    }

    private fun initializeLogger() {
        Logger.shared().platformService = LoggerPlatformServiceImpl(LOG_GROUP)

        if (BuildConfig.DEBUG) {
            Logger.shared().setLevel(LoggerLevel.VERBOSE)
        } else {
            Logger.shared().setLevel(LoggerLevel.ERROR)
        }
    }

    private fun initializeSharedData() {
        SharedData.shared().platformService = SharedDataPlatformServiceImpl(applicationContext)
    }

    private fun initializeFileHelper() {
        FileHelper.shared().platformService = FileHelperPlatformServiceImpl()
    }

    private fun initializeHttpClient() {
        // http client (uncomment to use platform http client)
        HttpClient.shared().platformService = HttpClientPlatformServiceImpl()
    }

    private fun initializeStrictMode() {
        if (BuildConfig.DEBUG) {
            StrictMode.setThreadPolicy(
                StrictMode.ThreadPolicy.Builder()
                    .detectAll()
                    .penaltyLog()
                    .build()
            )

            StrictMode.setVmPolicy(
                StrictMode.VmPolicy.Builder()
                    .detectLeakedSqlLiteObjects()
                    .detectLeakedClosableObjects()
                    .penaltyLog()
                    .build()
            )
        }
    }

    private fun initializeHttpServer() {
        Logger.i("[Application : initializeHttpServer]")

        if (appData.isInstantApp) {
            Logger.i("[Application : initializeHttpServer] Is instant app")
            return
        }

        val basePath = ApplicationCore.shared().initializationData.basePath
        val targetFolder = FileHelper.join(basePath, "webapp")

        FileHelper.removeDir(targetFolder)

        val assetManager: AssetManager = assets
        assetManager.copyAssetFolder(
            "webapp",
            targetFolder
        )

        val config = HttpServerConfig(9090, "")
        HttpServer.shared().initialize(config)
        startHttpServer()

        Logger.i("[Application : initializeHttpServer] Server: " + HttpServer.shared().socketAddress)
    }

    private fun startHttpServer() {
        Logger.i("[Application : startHttpServer]")

        HttpServer.shared().stop()
        HttpServer.shared().start()
    }

    private fun stopHttpServer() {
        Logger.i("[Application : stopHttpServer]")

        HttpServer.shared().stop()
    }

    companion object {
        private const val LOG_GROUP = "SAMPLE"

        lateinit var instance: Application
            private set
    }
}
